zilliqa/src/consensus.rs

```diff
@@ -515,18 +500,26 @@ impl Consensus {
 
         match self.check_block(&block, during_sync) {
             Ok(()) => {}
-            Err(e) => {
-                if let Some(e) = e.downcast_ref::<MissingBlockError>() {
-                    info!(?e, "missing block when checking block proposal - try and request the parent from the network: {}", block.header.number.saturating_sub(1));
-
+            Err((e, temporary)) => {
+                // If this block could become valid in the future, buffer it.
+                if temporary {
                     let random_peer = self.get_random_other_peer();
-                    self.block_store
-                        .request_blocks(random_peer, block.header.number.saturating_sub(1))?;
-                    return Ok(None);
-                } else {
-                    warn!(?e, "invalid block proposal received!");
-                    return Ok(None);
+                    self.block_store.buffer_proposal(
+                        Proposal::from_parts_with_hashes(
+                            block,
+                            transactions
+                                .into_iter()
+                                .map(|tx| {
+                                    let hash = tx.calculate_hash();
+                                    (tx, hash)
+                                })
+                                .collect(),
+                        ),
+                        random_peer,
```

Comment:
Note that if the proposal buffer contains more than its max configured size, we will in fact not buffer the proposal - is that OK (or at least, given that the buffer is LRU, we will throw a proposal away, which may have been the result of a previous call from here). Happy with that?

Reply:
I will be happy with that once we have retries...
https://github.com/Zilliqa/zq2/issues/1096

I've added a comment to point to this.

Reply:
:-) - sounds good to me; not hitting the resolve button because I can't see the comment (yet), but feel free to click the resolve button if you are happy.